src/main/java/org/terasology/AcceptanceTestEnvironment.java

```
@@ -0,0 +1,98 @@
+ * {@link #testDuringReplay()} and {@link #testOnReplayEnd()} . If it is not desired to test something in one or more of
+ * the replay stages, it is possible to leave the methods in black.
+ * <p>
+ * After the implementation of the abstract methods, it is necessary to write a test method that calls {@link #runTest(String, boolean)} .
+ */
+public abstract class AcceptanceTestEnvironment extends ReplayTestingEnvironment {
+
+    private String recordingTitle;
+    private boolean isHeadless;
+
+    /** To test the replay while it is executing, it is necessary to create a thread that will run the replay. */
+    private Thread replayThread = new Thread() {
+
+        @Override
+        public void run() {
+            try {
+                String replayTitle = recordingTitle;
+                AcceptanceTestEnvironment.super.openReplay(replayTitle, isHeadless);
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+    };
+
+    @After
+    public void closeReplay() throws Exception {
+        super.getHost().shutdown();
+        GameThread.reset();
+        replayThread.join();
+    }
+
+    /**
+     * Executes the replay test.
+     * @param replayTitle the title of the replay.
+     * @param headless if the engine should be headless.
+     */
+    protected void runTest(String replayTitle, boolean headless) {
+        this.isHeadless = headless;
+        this.recordingTitle = replayTitle;
+        replayThread.start();
+        try {
+            waitUntil(() -> (isInitialised() && getRecordAndReplayStatus() == RecordAndReplayStatus.REPLAYING));
+            testOnReplayStart();
+            testDuringReplay();
+            waitUntil(() -> getRecordAndReplayStatus() == RecordAndReplayStatus.REPLAY_FINISHED);
+            testOnReplayEnd();
+        } catch (Exception e) {
+            e.printStackTrace();
```

Same as above. The problem here is even more glaring since you catch all exceptions that may be thrown by the actual testing (`testXYZ`) methods, which is a big no-no.